### PR TITLE
Feature cu hubspot unsubscribe webhook

### PR DESCRIPTION
**Pull request by cupshir**
This is the initial relase of the hubspot webhook. It includes the migration to create the new hubspot contact table.

Even though we are not going live yet, I would like to get this merged into dev so that I can test out deploying to staging to ensure migration works before going to production.

This is a big merge, any time spent reviewing the code is appreciated!!!

### CODE REVIEW

File: church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs [hunk 1]
```diff
@@ -0,0 +1,1093 @@
...
+                eventsToProcess.Add( singleEvent );
+            }
+
+            // loop through events and process
+            foreach ( JObject eventItem in eventsToProcess )
+            {
+                if ( !JsonContainsKey( eventItem, "subscriptionType" ) || !JsonContainsKey( eventItem, "objectId" ) )
+                {
+                    // not a valid event, skip to next one
+                    continue;
```
Comment (jhawkzz) [inline church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs]: I am not personally a fan of continue (and only cautiously use break). I prefer to wrap what it should do inside the if. This is personal preference.

File: church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs [hunk 2]
```diff
@@ -0,0 +1,1093 @@
...
+        /// <summary>
+        /// Process events from HubSpot Webhook
+        /// </summary>
+        /// <param name="rawEvents"></param>
+        /// <returns></returns>
+        public static async Task ProcessEvents( string rawEvents )
+        {
+            // parse events into JArray
+            JArray eventsToProcess = new JArray();
+            if ( rawEvents.Substring( 0, 1 ) == "[" )
```
Comment (jhawkzz) [inline church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs]: If rawEvents is null, this will throw an exception.

File: church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs [hunk 3]
```diff
@@ -0,0 +1,1093 @@
...
+        /// </summary>
+        /// <param name="objectToProcess"></param>
+        /// <returns></returns>
+        private static async Task ProcessContactCreation( JObject objectToProcess )
+        {
+            RockContext rockContext = new RockContext();            
+            Service<HubSpotContact> hubSpotContactService = new Service<HubSpotContact>( rockContext );
+
+            // check for existing contact
+            int hubSpotObjectId = ( int ) objectToProcess["objectId"];
```
Comment (jhawkzz) [inline church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs]: If `objectToProcess` is null, this will crash. (If it's impossible for a null argument to be passed in, don't worry about this.)

File: church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs [hunk 4]
```diff
@@ -0,0 +1,1093 @@
...
+
+        /// <summary>
+        /// Process contact deletion event
+        /// </summary>
+        /// <param name="objectToProcess"></param>
+        /// <returns></returns>
+        private static async Task ProcessContactDeletion( JObject objectToProcess )
+        {
+            // lookup contact info to see if this was a delete or a merge
+            int hubSpotObjectId = ( int ) objectToProcess["objectId"];
```
Comment (jhawkzz) [inline church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs]: If `objectToProcess` is null this will crash. If its impossible for null to be passed, this is fine.

File: church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs [hunk 5]
```diff
@@ -0,0 +1,1093 @@
...
+
+        /// <summary>
+        /// Process property change event from HubSpot
+        /// </summary>
+        /// <param name="hubSpotContact"></param>
+        /// <param name="objectToProcess"></param>
+        private static async Task ProcessPropertyChange( JObject objectToProcess )
+        {
+            // ensure the event has required keys
+            if ( !JsonContainsKey( objectToProcess, "propertyName") || !JsonContainsKey( objectToProcess, "propertyValue" ) )
```
Comment (jhawkzz) [inline church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs]: If `objectToProcess` is null, this will crash. If it's impossible for null to be passed, this is fine.

File: church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs [hunk 6]
```diff
@@ -0,0 +1,1093 @@
...
+                {
+                    // HubSpot missing rock_id, add to contact in HubSpot
+                    Dictionary<string, string> propertiesToUpdate = new Dictionary<string, string>
+                    {
+                        { "rock_id", hubSpotContact.PersonAliasId.ToString() }
+                    };
+
+                    HttpResponseMessage updateContactApiResponse = await UpdateContactInHubSpot( hubSpotObjectId, propertiesToUpdate );
+
+                    if ( updateContactApiResponse.StatusCode != HttpStatusCode.NoContent )
```
Comment (jhawkzz) [inline church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs]: I wouldn't leave an empty if statement here. I'd comment it out if you need it for a future TODO reference.

File: church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs [hunk 7]
```diff
@@ -0,0 +1,1093 @@
...
+                case "address":
+                case "city":
+                case "state":
+                case "zip":
+                case "country":
+                case "hs_email_optout":
+                {
+                    await UpdatePersonInRock( hubSpotContact, propertyName, propertyValue );
+                }
+                    break;
```
Comment (jhawkzz) [inline church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs]: `break` should be within the {}s right below the `await`.

File: church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs [hunk 8]
```diff
@@ -0,0 +1,1093 @@
...
+                        if ( excludeStaff == false && attributeIdValue.IsNotNullOrWhitespace() )
+                        {
+                            if ( int.TryParse( attributeIdValue, out int attributeId ) )
+                            {
+                                UpdatePersonAttributeValueInRock( hubSpotContact, attributeId, attributeFieldType, propertyValue );
+                            }
+                        }
+                    }
+                }
+                    break;
```
Comment (jhawkzz) [inline church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs]: `break` should be within the {}s of the `default` case.

File: church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs [hunk 9]
```diff
@@ -0,0 +1,1093 @@
...
+                {
+                    await UpdatePersonInRock( hubSpotContact, propertyName, propertyValue );
+                }
+                    break;
+                default:
+                {
+                    // if no person field was specified, then this is a custom attribute
+                    // load the hubSpotPropertyMap to determine if we have a mapped attraibute
+                    var hubSpotPropertyMap = DefinedTypeCache.Read( church.ccv.Utility.SystemGuids.DefinedType.HUBSPOT_PROPERTY_MAP.AsGuid() );
+                    var hubSpotProperty = hubSpotPropertyMap.DefinedValues.Where( a => a.Value == propertyName ).SingleOrDefault();
```
Comment (jhawkzz) [inline church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs]: `SingleOrDefault` will throw an exception if there are ever multiple values for this DefinedValue. If you're 100% positive this can't happen, this is fine. Otherwise use `FirstOrDefault`

File: church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs [hunk 10]
```diff
@@ -0,0 +1,1093 @@
...
+            if ( hubSpotContacts.Count == 0 )
+            {
+                // nothing found, nothing to do
+                return;
+            }
+
+            // remove each contact
+            foreach ( HubSpotContact hubSpotContact in hubSpotContacts )
+            {
+                hubSpotContactService.Delete( hubSpotContact );
```
Comment (jhawkzz) [inline church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs]: Create, save and destroy the `rockContext` within this iterator. Otherwise you risk bloating the context and timing out.

File: church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs [hunk 11]
```diff
@@ -0,0 +1,1093 @@
...
+
+                // update BirthYear
+                if ( birthDateValues.Length == 3 )
+                {
+                    if ( int.TryParse( birthDateValues[2], out int birthYear ) )
+                    {
+                        newPerson.BirthYear = birthYear;
+                    }
+                }
+            }
```
Comment (jhawkzz) [inline church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs]: 1. Did you consider 4 digit birth years vs 2? (Just asking to make sure you considered it)
2. This same code is used below on 685. I would put this in a private utility helper function within this same code file.

File: church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs [hunk 12]
```diff
@@ -0,0 +1,1093 @@
...
+                    // invalid number, skip update
+                    return;
+                }
+
+                // default to mobile phone type
+                PhoneNumberService phoneNumberService = new PhoneNumberService( rockContext );
+                DefinedValueCache phoneType = DefinedValueCache.Read( Rock.SystemGuid.DefinedValue.PERSON_PHONE_TYPE_MOBILE.AsGuid() );
+                if ( personField == "phone" )
+                {
+                    phoneType = DefinedValueCache.Read( Rock.SystemGuid.DefinedValue.PERSON_PHONE_TYPE_HOME.AsGuid() );
```
Comment (jhawkzz) [inline church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs]: I think you meant to use `PERSON_PHONE_TYPE_MOBILE` here. The comment above (as well as the phoneType you load) is MOBILE. I could be wrong but just double checking.

File: church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs [hunk 13]
```diff
@@ -0,0 +1,1093 @@
...
+
+                // update the number
+                phoneNumber.Number = propertyValue;
+
+                if ( phoneNumber.NumberTypeValueId == phoneType.Id )
+                {
+                    // ensure sms is enabled if its a mobile number
+                    phoneNumber.IsMessagingEnabled = true;
+                }
+            }
```
Comment (jhawkzz) [inline church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs]: You may want to evaluate using `Person.UpdatePhoneNumber` in Person.cs (line 3249). I remember there being some gotchas with Phone Numbers, and this solved them all.

File: church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs [hunk 14]
```diff
@@ -0,0 +1,1093 @@
...
+                    };
+
+                    groupLocationService.Add( homeAddress );
+                }
+                else
+                {
+                    if ( homeAddress.Location.Guid == newAddressLocation.Guid )
+                    {
+                        // locations are the same, skip update
+                        return;
```
Comment (jhawkzz) [inline church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs]: I believe if this return is executed, no changes to the person will be saved. The only `rockContext.saveChanges` call is at the bottom of this function, so this would early-exit resulting in nothing being updated.

As a side note, I try to avoid returning from functions within nested logic; it can be hard to follow what's going on.

File: church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs (hunk 2 above)
Comment (cupshir) [reply church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs]: Added a null / whitespace check before the Substring.

File: church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs (hunk 3 above)
Comment (cupshir) [reply church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs]: I am checking for the existence of objectId in the method that calls this one so it should always exist by this point. I put the check in the calling so that I didnt have to include a check in each of the possible processing methods.

File: church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs (hunk 4 above)
Comment (cupshir) [reply church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs]: I am checking for the existence of objectId in the method that calls this one so it should always exist by this point. I put the check in the calling so that I didnt have to include a check in each of the possible processing methods.

File: church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs (hunk 5 above)
Comment (cupshir) [reply church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs]: I am checking for the existence of objectId in the method that calls this one so it should always exist by this point. I put the check in the calling so that I didnt have to include a check in each of the possible processing methods.

File: church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs (hunk 6 above)
Comment (cupshir) [reply church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs]: removed.

File: church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs (hunk 7 above)
Comment (cupshir) [reply church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs]: great catch, moved to correct location!

File: church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs (hunk 8 above)
Comment (cupshir) [reply church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs]: great catch, moved to correct location!

File: church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs (hunk 12 above)
Comment (cupshir) [reply church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs]: hubspot sends over "phone" when the number is a home number. If they send a home number, we want to update the home number in rock.

File: church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs (hunk 10 above)
Comment (cupshir) [reply church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs]: I updated it to use a 2nd context that is created, saved, and destroyed within the iterator.

File: church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs (hunk 11 above)
Comment (cupshir) [reply church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs]: I moved the update birthday code into a helper function. Also, I updated so that it will only update birth year if 4 digits are provided.  Good catch, if I passed 79, rock was setting date to 0079.

File: church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs (hunk 13 above)
Comment (cupshir) [reply church.ccv.CCVCore/HubSpot/Util/HubSpotContactService.cs]: Refactored phone number updates to use Person.UpdatePhoneNumber
